prow/cmd/initupload/main.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"flag"
+	"io/ioutil"
+	"path"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/option"
+
+	"k8s.io/test-infra/prow/pjutil"
+	"k8s.io/test-infra/prow/pod-utils/clone"
+	"k8s.io/test-infra/prow/pod-utils/gcs"
+)
+
+var (
+	cloneLog = flag.String("clone-log", "", "path to the output file for the cloning step")
+
+	subDir = flag.String("sub-dir", "", "Optional sub-directory of the job's path to which artifacts are uploaded")
+
+	pathStrategy = flag.String("path-strategy", "explicit", "how to encode org and repo into GCS paths")
+	defaultOrg   = flag.String("default-org", "", "optional default org for GCS path encoding")
+	defaultRepo  = flag.String("default-repo", "", "optional default repo for GCS path encoding")
+
+	gcsBucket          = flag.String("gcs-bucket", "", "GCS bucket to upload into")
+	gceCredentialsFile = flag.String("gcs-credentials-file", "", "file where Google Cloud authentication credentials are stored")
+	dryRun             = flag.Bool("dry-run", true, "do not interact with GCS")
+)
+
+func main() {
```

Comment:
Can we unit test this please? Lots of interesting stuff happening in here! :)

Reply:
Added some but having a hard time figuring out best places to test this... it's a bunch of non-trivial interaction with the filesystem followed by trivial actions in `go`. Happy to add more tests as you see fit